_source/_posts/2022-04-21-jhipster-k8s-azure.md

```diff
@@ -0,0 +1,669 @@
+# in gateway
+./gradlew bootJar -Pprod jib -Djib.to.image=$DOCKER_REPO_NAME/gateway
+```
+
+To briefly explain what's happening here, take a look at the blog service Kubernetes descriptor file. It defines a container named `blog-app` that uses the docker image `andrewcarterhughes/blog`, which is my docker repository name and the blog image.
+
+`k8s/blog-k8s/blog-deployment.yml`
+
+```yml
+containers:
+  - name: blog-app
+    image: andrewcarterhughes/blog
+```
+
+Thus in the `k8s` directory, each service (store, blog, gateway, and registry) defines a container and a docker image to be run in that container, along with a whole lot of configuration (which is really a lot of what JHipster is bootstrapping for you). The registry does not have a project folder because it uses a stock image that can be pulled directly from the JHipster docker repository.
+
+One nice feature of Kubernetes is the ability to define [init containers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/). These are containers that run before the main container and can be used to create or wait for necessary resources like databases. I noticed while I was debugging things in this app that a lot of the errors happened in the init containers. It's helpful to know this because if you try and inspect the main container log nothing will be there because the container hasn't even started you. You have to check the log for the init container that failed. The Kubernetes management tools that I mention below really come in handy for this.
+
+#### Deploy the Microservice to Azure AKS
+
+You can manage a Kubernetes service purely with `kubectl`. However, there are some pretty helpful tools for monitoring and logging. Both [k9s](https://github.com/derailed/k9s) and [Kubernetes Lens](https://k8slens.dev/) are great. I recommend installing one or both of these and using them to inspect and monitor your Kubernetes services. They are especially helpful when things go wrong (not that things ever go wrong, I wouldn't know anything about that, I just heard about it from friends, I swear). Kubernetes Lens is a full-on desktop app that describes itself as a Kubernetes IDE. In comparison, k9s is a lighter-weight, text-based tool. 
+
+Open a Bash shell and navigate to the `k8s` subdirectory of the project.
+
+Run the microservice with: 
+
+```bash
+./kubectl-apply.sh -f
+```
+
+If you open this file, you'll see that it creates the namespace and applies the project files. If you do this manually, it's important that the namespace is created first and that the registry is run before the other services.
+
+```bash
+...
+suffix=k8s
+kubectl apply -f namespace.yml
+kubectl apply -f registry-${suffix}/
+kubectl apply -f blog-${suffix}/
+kubectl apply -f gateway-${suffix}/
+kubectl apply -f store-${suffix}/
+...
+```
+
+You can check on your pods with the following command. Everything was up and running almost immediately for me on AKS.
```

Comment:
I don't know if 3m is _almost immediately_. 😜

Here's what I saw:

```
NAME                                  READY   STATUS    RESTARTS   AGE
blog-77965d99f6-qr6bs                 0/1     Running   0          94s
blog-neo4j-0                          1/1     Running   0          92s
gateway-5c95c95fdf-9dcgp              0/1     Running   0          89s
gateway-postgresql-647476b4d5-h5j5v   1/1     Running   0          87s
jhipster-registry-0                   1/1     Running   0          96s
jhipster-registry-1                   1/1     Running   0          83s
store-7778fcfc59-x2xxd                1/1     Running   0          84s
```

Reply:
Mine was faster than that. But I was on the upgraded plan.